### PR TITLE
Va/gpf/conditional kernel

### PR DESCRIPTION
**Pull request by thevincentadam**
### Conditioned kernel

New feature: A wrapper around kernel to condition on noiseless observations.

### Minimal working example

`from gpflow.kernels import RBF, Conditioned`
`D, N, Nc = 1, 200, 2`
`Xc = np.random.uniform(low=-1, high=1., size=(Nc, D))`
`Yc = np.random.uniform(low=-1, high=1., size=(Nc, D))`
`k = Conditioned(base=RBF(variance=1., lengthscale=.2), xc=Xc, yc=Yc)`

### CODE REVIEW

**Comment by st--** [comment]
@thevincentadam could you please
* clarify what the intent behind this PR is (i.e. what would be use-cases? what would other people use this for? why should GPflow carry the maintenance burden?)
* add test coverage (happy to help figure out what kind of tests might be useful)?

File: gpflow/kernels/conditioned.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
+from typing import Optional
+import tensorflow as tf
+from .base import Kernel
+from ..config import default_float, default_jitter
+
+
+class Conditioned(Kernel):
```
Comment (awav) [inline gpflow/kernels/conditioned.py]: One of the criteria for merging PRs is test cases for new features. Could you add tests?

File: gpflow/kernels/conditioned.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+            raise TypeError("Conditioned requires a Kernel object as the `base`")
+
+        super().__init__()
+        self.base = base
+        self.xc = xc
+        self.yc = yc
+        self.num_c = xc.shape[0]
+
+    @property
+    def Lc(self):
```
Comment (awav) [inline gpflow/kernels/conditioned.py]: The naming for this function can be improved. Can you come up with another name?

File: gpflow/kernels/conditioned.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+    From a covariance function k(.,.) and a conditioning dataset (Xc, Yc),
+    a new covariance function and a conditional mean function
+    are created as
+
+       kc(.,.) = k(.,.) - k(.,Xc)k(Xc, Xc)⁻¹k(Xc,.)
+       mc(.)   = k(.,Xc)k(Xc, Xc)⁻¹Yc
+    """
+
+    def __init__(self, base: Kernel, xc: tf.Tensor, yc: tf.Tensor):
```
Comment (awav) [inline gpflow/kernels/conditioned.py]: The `xc` and `yc` are very obscure names. Could you call them differently to reflect their meaning?

File: gpflow/kernels/conditioned.py (hunk 1 above)
Comment (thevincentadam) [reply gpflow/kernels/conditioned.py]: added one!

File: gpflow/kernels/conditioned.py (hunk 3 above)
Comment (thevincentadam) [reply gpflow/kernels/conditioned.py]: renamed!

File: gpflow/kernels/conditioned.py (hunk 2 above)
Comment (thevincentadam) [reply gpflow/kernels/conditioned.py]: renamed!

**Comment by markvdw** [comment]
I'm also curious to hear the answer to @st--'s questions. Although I do kindof see that this is cool, I worry that it's duplication of functionality in `GPR`, which would increase the maintenance burden without adding more functionality.

According to the Zen of Python... `There should be one-- and preferably only one --obvious way to do it.`

**Comment by st--** [comment]
Hi @thevincentadam, if you are still interested in working on this PR it'd be good if you could answer the questions above. Thanks!

**Comment by st--** [comment]
@vincentadam87 I'm closing this PR as it is out of date with develop and you haven't responded to the questions above yet; please feel free to re-open with updates on that!

**Pull request closed without merging**